File: antlropt/org/perses/antlr/pnf/AbstractPnfPass.java
package org.perses.antlr.pnf;

import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.Iterables;
import org.perses.antlr.ast.*;

import java.util.*;
import java.util.stream.Collectors;

public abstract class AbstractPnfPass {

  public abstract ImmutableRuleDefMap process(ImmutableRuleDefMap grammar);

  protected static List<RuleNameRegistry.RuleNameHandle> getSortedRuleNames(MutableRuleDefMap map) {
    return map.keySet().stream().sorted().collect(Collectors.toList());
  }

  protected static AbstractPersesRuleElement replaceGapWithRuleReference(
      PersesSequenceAst seq, Interval gap, RuleNameRegistry.RuleNameHandle gapReplacement) {
    final ImmutableList.Builder<AbstractPersesRuleElement> builder = ImmutableList.builder();
    for (int i = 0, size = seq.getChildCount(); i < size; ++i) {
      if (i == gap.leftInclusive) {
        builder.add(new PersesRuleReferenceAst(gapReplacement));
      }
      if (gap.leftInclusive <= i && i < gap.rightExclusive) {
        continue;
      }
      builder.add(seq.getChild(i));
    }
    return new PersesSequenceAst(builder.build());
  }

  protected static List<AbstractPersesRuleElement> constructAlternativeBlocksIfNecessary(
      Collection<AbstractPersesRuleElement> alternatives) {
    alternatives = flatternAltBlocks(alternatives);
    final int size = alternatives.size();
    Preconditions.checkArgument(size > 0);
    if (size == 1) {
      return ImmutableList.of(Iterables.getOnlyElement(alternatives));
    }
    final HashMap<String, AbstractPersesRuleElement> map = new HashMap<>();
    alternatives.forEach(alt -> map.put(alt.getSourceCode(), alt));
    return map.entrySet().stream()
        .map(Map.Entry::getValue)
        .collect(ImmutableList.toImmutableList());
  }

  protected static void inPlaceFlatternAltBlocks(List<AbstractPersesRuleElement> alternatives) {
    final ArrayList<AbstractPersesRuleElement> result = flatternAltBlocks(alternatives);
    alternatives.clear();
    alternatives.addAll(result);
  }

  protected static ArrayList<AbstractPersesRuleElement> flatternAltBlocks(
      Collection<AbstractPersesRuleElement> alternatives) {
    final ArrayList<AbstractPersesRuleElement> result = new ArrayList<>();
    for (AbstractPersesRuleElement alternative : alternatives) {
      flattern(alternative, result);
    }
    return result;
  }

  private static void flattern(
      AbstractPersesRuleElement element, ArrayList<AbstractPersesRuleElement> result) {
    if (element instanceof PersesAlternativeBlockAst) {
      final PersesAlternativeBlockAst block = (PersesAlternativeBlockAst) element;
      for (AbstractPersesRuleElement blockAlternative : block.getAlternatives()) {
        flattern(blockAlternative, result);
      }
    } else {
      result.add(element);
    }
  }

  protected static int countRulesThatCallsRuleOfName(
      MutableRuleDefMap grammar, RuleNameRegistry.RuleNameHandle ruleName) {
    RuleRefCounterAstVisitor counter = new RuleRefCounterAstVisitor(ruleName);
    counter.preorderGrammar(grammar.entries());
    return counter.getCount();
  }
}

File: antlropt/src/org/perses/antlr/pnf/AbstractPnfPass.kt
/*
 * Copyright (C) 2018-2024 University of Waterloo.
 *
 * This file is part of Perses.
 *
 * Perses is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3, or (at your option) any later version.
 *
 * Perses is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Perses; see the file LICENSE.  If not see <http://www.gnu.org/licenses/>.
 */
package org.perses.antlr.pnf

import org.perses.antlr.ast.AbstractPersesRuleElement
import org.perses.antlr.ast.PersesRuleReferenceAst
import org.perses.antlr.ast.PersesSequenceAst
import org.perses.antlr.ast.RuleNameRegistry.RuleNameHandle
import org.perses.antlr.pnf.AstUtil.replaceGapInSequence
import org.perses.util.Interval

abstract class AbstractPnfPass {

  abstract fun processGrammar(
    grammar: GrammarPair,
  ): GrammarPair

  companion object {
    @JvmStatic
    protected fun getSortedRuleNames(map: MutableGrammar): List<RuleNameHandle> {
      return map.ruleNameSequence().sorted().toList()
    }

    @JvmStatic
    protected fun replaceGapWithRuleReference(
      seq: PersesSequenceAst,
      gap: Interval,
      gapReplacement: RuleNameHandle,
    ): AbstractPersesRuleElement {
      val ruleRefAst = PersesRuleReferenceAst.create(gapReplacement)
      return replaceGapInSequence(seq, gap, ruleRefAst)
    }

    @JvmStatic
    protected fun countRulesThatCallsRuleOfName(
      grammar: MutableGrammar,
      ruleName: RuleNameHandle,
    ): Int {
      val counter = RuleRefCounterAstVisitor(ruleName)
      grammar.alternativeSequence().forEach {
        counter.preorder(it)
      }
      return counter.count
    }
  }
}
